ndinterp/src/grid.rs

```diff
@@ -10,19 +10,34 @@
 //!     y = f(x1, x2, x3...)
 //!
 use crate::interpolate::InterpolationError;
-use ndarray::Array1;
+use itertools::izip;
+use ndarray::{Array, ArrayView1, Dimension};
 
 // Make public the families of interpolation algorithms implemented for grids
 pub mod cubic;
 
+/// A grid is made of two components:
+///     A d-dimensional vector of 1-dimensional sorted vectors for the input points
+///     A d-dimensional array for the grid values of
 #[derive(Debug)]
-pub struct Grid {
-    /// A grid is made of two (1-dimensional) sorted arrays.
-    pub input: Array1<f64>,
-    pub values: Array1<f64>,
+pub struct Grid<D: Dimension> {
+    /// Arrays with the input vectors (x_i)
+    pub xgrid: Vec<Vec<f64>>,
+
+    /// Output points
+    pub values: Array<f64, D>,
+}
+
+/// A grid slice is always 1-Dimensional
+#[derive(Debug)]
+pub struct GridSlice<'a> {
+    /// A reference to one of the input vectors of the grid
+    pub x: &'a Vec<f64>,
```

Comment:
Depends on what you mean with “need”, that might also work (and be more correct?), but x is a reference to one of the quantities in the xgrid which are vectors hence the choice.

Reply:
If you need to use vector-specific operations, then yes, you need the reference to a vector.
But if you need to reference just a collection of points, the slice is more general, and it relies less on the underlying structure (while still be plenty of possible methods, https://doc.rust-lang.org/std/primitive.slice.html).

However, if it has to be one entire element (the full `x` coordinates) maybe it makes sense to keep as it is. I'd suggest instead to give it a name:
```rs
type XPoint = Vec<f64>;
```
(if you're able to find a better name, you're welcome). It's just an alias, for greater clarity, but here a newtype would be too much (I believe).

Reply:
Ah, no. I do need a vector because of the search function.

But it is not a point, it is actually the `x` grid for  \vec{y} = f(\vec{x}). I could call them `xgrid`, I decided to change it to `x` since it is not exposed to the user and only used internally (as opposed to the `xgrid` which needs to be filled by the user).

But I can use `xgrid` as well here if you prefer (but I'll do the change directly in #5)

Reply:
"better name" was referring to the name of the type alias I was proposing. `x` for the attribute is just fine.

The reason behind the alias is that an `XPoint` is isomorphic to a `Vec<f64>`, but not all `Vec<f64>` are `XPoint`s. It doesn't help with automatic check, because an alias is interchangeable with the aliased type, but it helps the developer reading it (to have help also from the compiler, we could use a [newtype](https://doc.rust-lang.org/rust-by-example/generics/new_types.html), but as I said it seems to me that it's not worth, at the moment)